gadopt/level_set_tools.py

```diff
@@ -766,32 +767,128 @@ class was initialised.
 
 
 def entrainment(
-    level_set: fd.Function, material_area: Number, entrainment_height: Number
-):
-    """Calculates entrainment diagnostic.
+    level_set: fd.Function, material_area: float, entrainment_height: float
+) -> float:
+    """Calculates the entrainment diagnostic.
 
-    Determines the proportion of a material that is located above a given height.
+    Determines the proportion of a material located above a given height.
 
     Args:
-        level_set:
-          A level-set Firedrake function.
-        material_area:
-          An integer or a float representing the total area occupied by a material.
-        entrainment_height:
-          An integer or a float representing the height above which the entrainment
-          diagnostic is determined.
+      level_set:
+        A Firedrake function for the level-set field
+      material_area:
+        A float representing the total area occupied by the target material
+      entrainment_height:
+        A float representing the height above which to calculate entrainment
 
     Returns:
-        A float corresponding to the calculated entrainment diagnostic.
+      A float corresponding to the material fraction above the target height
     """
-    mesh_coords = fd.SpatialCoordinate(level_set.function_space().mesh())
+    mesh_coords = node_coordinates(level_set)
     target_region = mesh_coords[1] >= entrainment_height
     material_entrained = fd.conditional(level_set < 0.5, 1, 0)
+    is_entrained = fd.conditional(target_region, material_entrained, 0)
 
-    return (
-        fd.assemble(fd.conditional(target_region, material_entrained, 0) * fd.dx)
-        / material_area
-    )
+    return fd.assemble(is_entrained * fd.dx) / material_area
+
+
+def min_max_height(
+    level_set: fd.Function, epsilon: float | fd.Function, *, side: int, mode: str
+) -> float:
+    """Calculates the maximum or minimum height of a material interface.
+
+    Args:
+      level_set:
+        A Firedrake function for the level set field
+      epsilon:
+        A float or Firedrake function denoting the thickness of the material interface
```

Comment:
I admit I'm new to level set code, but perhaps a tiny bit more detail specifying why this is needed.